homeassistant/components/dsmr/__init__.py

```diff
@@ -38,3 +52,77 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
 async def async_update_options(hass: HomeAssistant, entry: ConfigEntry) -> None:
     """Update options."""
     await hass.config_entries.async_reload(entry.entry_id)
+
+
+@callback
+def async_migrate_entity_entry(
+    config_entry: ConfigEntry, entity_entry: er.RegistryEntry
+) -> dict[str, Any] | None:
+    """Migrate DSMR entity entries.
+
+    - Migrates unique ID for sensors based on entity description name to key.
+    """
+
+    # Replace names with keys in unique ID
+    for old, new in (
+        ("Power Consumption", "current_electricity_usage"),
+        ("Power Production", "current_electricity_delivery"),
+        ("Power Tariff", "electricity_active_tariff"),
+        ("Energy Consumption (tarif 1)", "electricity_used_tariff_1"),
+        ("Energy Consumption (tarif 2)", "electricity_used_tariff_2"),
+        ("Energy Production (tarif 1)", "electricity_delivered_tariff_1"),
+        ("Energy Production (tarif 2)", "electricity_delivered_tariff_2"),
+        ("Power Consumption Phase L1", "instantaneous_active_power_l1_positive"),
+        ("Power Consumption Phase L2", "instantaneous_active_power_l2_positive"),
+        ("Power Consumption Phase L3", "instantaneous_active_power_l3_positive"),
+        ("Power Production Phase L1", "instantaneous_active_power_l1_negative"),
+        ("Power Production Phase L2", "instantaneous_active_power_l2_negative"),
+        ("Power Production Phase L3", "instantaneous_active_power_l3_negative"),
+        ("Short Power Failure Count", "short_power_failure_count"),
+        ("Long Power Failure Count", "long_power_failure_count"),
+        ("Voltage Sags Phase L1", "voltage_sag_l1_count"),
+        ("Voltage Sags Phase L2", "voltage_sag_l2_count"),
+        ("Voltage Sags Phase L3", "voltage_sag_l3_count"),
+        ("Voltage Swells Phase L1", "voltage_swell_l1_count"),
+        ("Voltage Swells Phase L2", "voltage_swell_l2_count"),
+        ("Voltage Swells Phase L3", "voltage_swell_l3_count"),
+        ("Voltage Phase L1", "instantaneous_voltage_l1"),
+        ("Voltage Phase L2", "instantaneous_voltage_l2"),
+        ("Voltage Phase L3", "instantaneous_voltage_l3"),
+        ("Current Phase L1", "instantaneous_current_l1"),
+        ("Current Phase L2", "instantaneous_current_l2"),
+        ("Current Phase L3", "instantaneous_current_l3"),
+        ("Max power per phase", "belgium_max_power_per_phase"),
+        ("Max current per phase", "belgium_max_current_per_phase"),
+        ("Energy Consumption (total)", "electricity_imported_total"),
+        ("Energy Production (total)", "electricity_exported_total"),
+    ):
+        old = old.replace(" ", "_")
+        if entity_entry.unique_id.endswith(old):
```

Comment:
This is quite expensive at startup and is only needed the first time. Would there be a faster way to check if a unique ID needs migrating? Or maybe we store `"migrated": True` in the config entry data and only migrate the very first time.

Reply:
This is how our migration helper works. We have versions on config entries, but are not allowed to use these either.

Using an marker like you suggest is just as bad.

Reply:
How long do we expect to have `unique_id` migrations in place? Is there a supported period like 6 months or are they supposed to stay forever?

Reply:
Why not just match the key with what the old unique id ended up being. Then you dont need any migration. You already change to grab data using the new obis_reference, so the value for key is only home assistant internal right?

Reply:
> Why not just match the key with what the old unique id ended up being

Because it's rubbish.

Reply:
Not sure i really care if the unique id is rubbish. Its not visible to user. But if you feel strongly about it :), as long as we get rid of the conversion at some poin.

Reply:
I do, I like to keep stuff clean. And the part comment on / discussed now, is about the migration tools we provide to integration developers to deal with these things.

That is just odd. If we provide the tools, we should be able to use them. If those are not good, that should be fixed or improved but should not limit integration improvements (regardless of it being user-facing or not).

Reply:
As discussed in chat/VR, this is what we currently offer/point others into on how to solve this issue/do the migration. As we don't have an alternative at this point, we move forward with this method right now.